lib/Semantic_check.ml

```diff
@@ -497,7 +304,221 @@ and inferred_unsizedtype_of_indexed loc ut typed_indexl =
+         generated quantities block or user-defined functions with names \
+         ending in _rng."
+      , loc )
+    |> Or_error.of_exn
+  else Result.Ok (id, es)
+
+(* Regular function application *)
+let semantic_check_fn_normal ~loc (id, es) =
+  match Symbol_table.look vm id.name with
+  | Some (_, Mir.UFun (_, Void)) ->
+      SemanticError
+        ( Format.sprintf
+            "A returning function was expected but a non-returning function \
+             '%s' was supplied."
+            id.name
+        , loc )
+      |> Or_error.of_exn
+  | Some (_, UFun (arg_types, _))
+    when not (check_compatible_arguments_mod_conv id.name arg_types es) ->
+      let msg =
+        Format.sprintf
+          "Ill-typed arguments supplied to function '%s'. Available \
+           signatures: %s\n\
+           Instead supplied arguments of incompatible type: %s."
+          id.name
+          (Stan_math_signatures.pretty_print_all_math_lib_fn_sigs id.name)
+          (List.map es ~f:type_of_expr_typed |> pretty_print_unsizedtypes)
+      in
+      SemanticError (msg, loc) |> Or_error.of_exn
+  | Some (_, UFun (_, ReturnType ut)) ->
+      Result.Ok
+        { expr_typed= FunApp (Mir.UserDefined, id, es)
+        ; expr_typed_ad_level= lub_ad_e es
+        ; expr_typed_type= ut
+        ; expr_typed_loc= loc }
+  | Some _ ->
+      (* Check that Funaps are actually functions *)
+      SemanticError
+        ( Format.sprintf
+            "A returning function was expected but a non-function value '%s' \
+             was supplied."
+            id.name
+        , loc )
+      |> Or_error.of_exn
+  | None ->
+      SemanticError
+        ( Format.sprintf
+            "A returning function was expected but an undeclared identifier \
+             '%s' was supplied."
+            id.name
+        , loc )
+      |> Or_error.of_exn
+
+(** Stan-Math function application 
+*)
+let semantic_check_fn_stan_math ~loc (id, es) =
+  match Stan_math_signatures.stan_math_returntype id.name es with
+  | Some Void ->
+      id.name
+      |> Format.sprintf
+           "A returning function was expected but a non-returning function \
+            '%s' was supplied."
+      |> Or_error.error_string
+  | Some (ReturnType ut) ->
+      Result.Ok
+        { expr_typed= FunApp (Mir.StanLib, id, es)
+        ; expr_typed_ad_level= lub_ad_e es
+        ; expr_typed_type= ut
+        ; expr_typed_loc= loc }
+  | _ ->
+      let err =
+        Format.sprintf
+          "Ill-typed arguments supplied to function '%s'. Available \
+           signatures: %s\n\
+           Instead supplied arguments of incompatible type: %s."
+          id.name
+          (Stan_math_signatures.pretty_print_all_math_lib_fn_sigs id.name)
+          (List.map es ~f:type_of_expr_typed |> pretty_print_unsizedtypes)
+      in
+      Or_error.error_string err
+
+let fn_kind_from_identifier id =
+  if Stan_math_signatures.is_stan_math_function_name id.name then Mir.StanLib
```

Comment:
All of the names in that module are super long and redundant especially paired with the module name... Can we go back to opening it?